src/core/toga/sources/base.py

```diff
@@ -0,0 +1,71 @@
+
+
+class Datum:
+    def __init__(self, value=None, icon=None):
+        self.value = value
+        self.icon = icon
+
+    def __str__(self):
+        if self.value is None:
+            return ''
+        return str(self.value)
+
+    @property
+    def icon(self):
+        return self._icon
+
+    @icon.setter
+    def icon(self, path_or_icon):
+        from toga import Icon
+        if path_or_icon:
+            self._icon = Icon.load(path_or_icon)
+        else:
+            self._icon = None
+
+
+class Row:
+    def __init__(self, **data):
+        for name, value in data.items():
+            setattr(self, name, value)
+
+
+class Node(Row):
+    def __init__(self, **data):
+        super().__init__(**data)
+        self.parent = None
+        self._children = None
+
+
+class Source:
+    def __init__(self):
+        self._listeners = []
+
+    @property
+    def listeners(self) -> list:
+        """ The listeners of this data source.
+        Listeners can be ``callable`` or :obj:``toga.Widget``.
+
+        Returns:
+            A list of objects that are listening for data change.
+        """
+        return self._listeners
+
+    def add_listener(self, listener):
+        """
+        Args:
+            listener: ``callable`` or :obj:``toga.Widget`
+        """
+        self._listeners.append(listener)
+
+    def remove_listener(self, listener):
+        self._listeners.remove(listener)
+
+    def _notify(self, notification, *args, **kwargs):
+        """Invoke a notification function on all listeners that are subscribed to this data source."""
+        for listener in self._listeners:
+            getattr(listener, notification)(*args, **kwargs)
+
+
+def to_accessor(heading):
+    "Convert a human-readable heading into a data attribute accessor"
+    return heading.lower().replace(' ', '_')
```

Comment:
This seems _super_ brittle. Do we already have checking for other special characters in headings? What if I put in a newline, or emoji?

I realize this was already an existing bug, if we aren't doing that level of checking, but this change brings it to light.

Reply:
It wasn't an existing bug - previously, we weren't doing attribute lookup. 

Yes it *is* brittle as currently written - this was a very quick implementation. Before the final version, I'd expect to strip out any non-[\w+] characters, and so on.

It's also worth noting that you can override the accessors if you've got a heading that is ambiguous, repeated, or doesn't convert easily/well.